File: leetcode/python/1351_count_negative_numbers.py
class Solution:
    def countNegatives(self, grid: list[list[int]]) -> int:
        res: int = 0
        for arr in grid:
            l = 0
            r = len(arr)
            while l < r:
                m = (l+r) // 2
                if arr[m] < 0:
                    r = m
                else:
                    l = m+1
            res += len(arr) - l
        return res

s = Solution()
print(s.countNegatives([[4,3,2,-1],[3,2,1,-1],[1,1,-1,-2],[-1,-1,-2,-3]]))
print(s.countNegatives([[3,2], [1,0]]))

File: leetcode/rust/src/1351_count_negative_numbers.rs
pub fn count_negatives(grid: Vec<Vec<i32>>) -> i32 {
    grid.iter()
        .map(
            // do binary search on reversed row
            |row| (row.len() - row.partition_point(|&x| x >= 0)) as i32
        )
        .sum()
}